### ISSUE TITLE
when and how often is config callback called exactly?

### ISSUE THREAD

**Issue by haraldox** [lhorie/mithril.js#681]
Hi Leo & fellow contributors, much praise for you: I'm loving the minimalistic approach of mithril.js and prefer it much rather to React!

Since I feel myself starting to push the boundaries of minimalism while trying to integrate [iScroll](http://iscrolljs.com) with Mithril (similar to [react-iscroll](https://github.com/schovi/react-iscroll)) I've started to play around with the `config` attribute of `m()`, see https://lhorie.github.io/mithril/mithril.html#the-config-attribute

When and how often exactly is the supplied callback executed? For example, if you take the standard example on https://lhorie.github.io/mithril/getting-started.html and add a config attribute to any `m()` call in the view (e.g. `m("html", ...` or `m("td", ...`) it is executed once for every triggered `onchange` event of the input field or `onclick` of the "Add" button...

If you replace `onchange` with `oninput`, the config callback (for whatever unrelated virtual element it is defined) is triggered for each keystroke. As per documentation:

> Also note that the config callback only runs after a rendering lifecycle is done.

I've come to understand `m.render()` actually touches the DOM, whereas `m.redraw` only recomputes changes ... Is the rendering lifecycle explained in more detail somewhere?

Thanks for shedding some light on this issue!

**Comment by barneycarroll** [comment]
> Is the rendering lifecycle explained in more detail somewhere?

@haraldox no, but it really should be. Rather than descending into the voluminous walls of text favored by other popular frameworks, Mithril could really do with an internal APIs documentation that addressed the implicit lifecycle (to be internally linked by relevant public APIs). Because Mithril does so much stuff implicitly (as opposed to say Backbone or React, which give you massive API surfaces and tell you to do it all yourself). Advanced Mithril development pretty much depends on an understanding of the redraw lifecycle AFAIC, but it's recondite stuff learnt by hard won experience ('just read the source' is not a serious option IMO).

To answer your question:

TLDR: configs execute at the last step of every redraw after all DOM modifications, if any, have taken place.

Long answer:

Every event handler (like `oninput`) bound by Mithril queues a redraw after execution, the equivalent of calling `m.redraw` manually.

The queue is resolved asynchronously [1] with a delay deemed appropriate by the browser (or within 16 milliseconds), so further redraw requests within that period have no effect by default.

When the call to redraw resolves, Mithril checks the redraw strategy — 'diff' by default [2] — and computes the top level component's view, including all subviews, recursively.

When the view is computed, Mithril takes a snapshot of it. It then goes through the build process — iterates over each node and works out what has changed compared to the node in the same relative positron of the previous snapshot, then modifies the attributes that have changed, or creates the element from scratch and inserts them into the document [3].

Next, every config function registered during the build process is executed.

[1] If `true` is passed to an `m.redraw` call, or if `m.mount` is being called, then the render function will not be queued and will execute immediately instead.

[2] You can change redraw strategy at any moment (Mithril will reset it to 'diff' after every redraw and set it to 'all' when a route change occurs): 'none' essentially means 'ignore the last redraw request'; 'all' means don't perform any diffing and build the whole view from scratch.

[3] There are all sorts of exceptions ;)

**Comment by lhorie** [comment]
Have you had a chance to look at this page?

http://lhorie.github.io/mithril/auto-redrawing.html

**Comment by haraldox** [comment]
first of all, many thanks to @barneycarroll for your detailed and lengthy explanation! I'm still playing around with `config` and wrapping my head around the lifecycle...

@lhorie I have. In fact I am reading all of the available documentation pages over and over in my quest to master Mithril's concepts. The documentation is pretty concise, yet covers many of the requirements of  a webapp framework. I must say I like it better and better the more I delve in.

I still have tons of questions and many problems , but at the moment I'm trying to solve them by myself, so I can come up with some concrete issues before posting again.

many thanks for your continued developments efforts! :+1: I think many users appreciate it :)

**Comment by haraldox** [comment]
I already have a short question about the redrawing lifecycle if I may: As I've stated above I am trying to integrate the iScroll library with Mithril ... I have studied http://lhorie.github.io/mithril/integration.html however I need to implement 'the other way around', i.e. [iScroll needs to be refreshed](http://iscrolljs.com/#refresh) when the model changes but only after mithril has auto-redrawn the corresponding view.

I'm not quite sure how:
a) if there is a callback or similar functionality to inform iScroll after the virtual child view elements of the iscroll's `#wrapper` div has been redrawn
b) when a virtual element's `config` function is called exactly with regard to DOM redrawing (before, after...)

**Comment by lhorie** [comment]
configs run after the DOM is updated. You can typically sync up 3rd party libraries from that config callback
